chain/src/main/scala/org/bitcoins/chain/blockchain/ChainHandler.scala

```diff
@@ -150,32 +150,100 @@ case class ChainHandler(
-      stopBlockOpt <- getHeadersAtHeight(stopHeight).map(_.headOption)
-      stopBlock = stopBlockOpt.getOrElse(
-        throw UnknownBlockHeight(s"Unknown header height ${stopHeight}"))
+      prevBlockHeaderOpt <- prevBlockHeaderOptF
+
+      headerOpt <- prevBlockHeaderOpt match {
+        case Some(_) =>
+          findNextHeader(prevBlockHeaderOpt, batchSize)
+        case None =>
+          val result = if (prevStopHash == DoubleSha256DigestBE.empty) {
+            for {
+              next <- findNextHeader(None, batchSize)
+            } yield next
+          } else {
+            Future.successful(None)
+          }
+          result
+      }
     } yield {
-      if (startHeight > stopHeight)
-        None
-      else
-        Some((startHeight, stopBlock.hashBE.flip))
+      headerOpt
     }
   }
 
+  /** Finds the next header in the chain. Uses chain work to break ties
+    * returning only the header in the chain with the most work
+    */
+  private def findNextHeader(
+      prevBlockHeaderOpt: Option[BlockHeaderDb],
+      batchSize: Int): Future[Option[(Int, DoubleSha256Digest)]] = {
+
+    val chainsF = prevBlockHeaderOpt match {
+      case None =>
+        blockHeaderDAO.getBlockchainsBetweenHeights(from = 0,
+                                                    to = batchSize - 1)
+      case Some(prevBlockHeader) =>
+        val inMemoryBlockchains = {
+          blockchains.filter(
+            _.exists(_.previousBlockHashBE == prevBlockHeader.hashBE))
+        }
+        if (inMemoryBlockchains.nonEmpty) {
+          Future.successful(inMemoryBlockchains)
+        } else {
+          blockHeaderDAO.getBlockchainsBetweenHeights(
+            from = prevBlockHeader.height,
+            to = prevBlockHeader.height + batchSize)
+        }
+    }
+
+    val startHeight = prevBlockHeaderOpt match {
+      case None => 0
+      case Some(prevBlockHeader) =>
+        prevBlockHeader.height + 1
+    }
+
+    for {
+      chains <- chainsF
+    } yield getBestChainAtHeight(startHeight = startHeight,
+                                 batchSize = batchSize,
+                                 blockchains = chains)
+  }
+
+  /** Given a vector of blockchains, this method finds the chain with the most chain work
+    * and then returns the given height and hash of the block header included in that chain
+    * This is useful for reorg situations where you aren't sure what header is included in a chain
+    * at a specific height
+    * @see https://github.com/bitcoin-s/bitcoin-s/issues/1919
+    */
+  private def getBestChainAtHeight(
+      startHeight: Int,
+      batchSize: Int,
+      blockchains: Vector[Blockchain]): Option[(Int, DoubleSha256Digest)] = {
+    //ok, we need to select the header that is contained in the chain
+    //with the most chain work
+    val targetHeight = startHeight + batchSize - 1
+    val mostWorkChainOpt = org.bitcoins.core
+      .seqUtil(blockchains)
+      .maxByOption(_.tip.chainWork)
```

Comment:
nit: you can `import org.bitcoins.core.seqUtil` then do `blockchains.maxByOption(..)`